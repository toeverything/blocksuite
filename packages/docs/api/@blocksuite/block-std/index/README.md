[**@blocksuite/block-std**](../../../@blocksuite/block-std/README.md)

***

[BlockSuite API Documentation](../../../README.md) / [@blocksuite/block-std](../README.md) / index

# index

## Classes

- [~~BlockService~~](classes/BlockService.md)
- [CommandManager](classes/CommandManager.md)
- [LifeCycleWatcher](classes/LifeCycleWatcher.md)

## Functions

- [BlockViewExtension](functions/BlockViewExtension.md)
- [ConfigExtensionFactory](functions/ConfigExtensionFactory.md)
- [FlavourExtension](functions/FlavourExtension.md)
- [KeymapExtension](functions/KeymapExtension.md)
- [WidgetViewExtension](functions/WidgetViewExtension.md)
