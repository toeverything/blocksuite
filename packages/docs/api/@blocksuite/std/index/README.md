[**BlockSuite API Documentation**](../../../README.md)

***

[BlockSuite API Documentation](../../../README.md) / [@blocksuite/std](../README.md) / index

# index

## Classes

- [~~BlockService~~](classes/BlockService.md)
- [CommandManager](classes/CommandManager.md)
- [LifeCycleWatcher](classes/LifeCycleWatcher.md)

## Functions

- [BlockViewExtension](functions/BlockViewExtension.md)
- [ConfigExtensionFactory](functions/ConfigExtensionFactory.md)
- [FlavourExtension](functions/FlavourExtension.md)
- [KeymapExtension](functions/KeymapExtension.md)
- [WidgetViewExtension](functions/WidgetViewExtension.md)
