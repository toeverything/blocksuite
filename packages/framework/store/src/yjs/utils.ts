import type { Doc as YDoc } from 'yjs';

export type SubdocEvent = {
  added: Set<YDoc>;
  loaded: Set<YDoc>;
  removed: Set<YDoc>;
};
