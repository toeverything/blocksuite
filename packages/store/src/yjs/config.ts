export type ProxyConfig = {
  readonly?: boolean;
  deep?: boolean;
};
