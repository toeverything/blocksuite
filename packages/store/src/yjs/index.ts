export * from './doc.js';
export * from './proxy.js';
export * from './utils.js';
