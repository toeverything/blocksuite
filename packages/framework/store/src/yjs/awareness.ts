import { Slot } from '@blocksuite/global/utils';
import { merge } from 'merge';
import type { Awareness as YAwareness } from 'y-protocols/awareness.js';

import type { Space } from '../store/space.js';
import type { Store } from '../store/store.js';

export interface UserInfo {
  name: string;
}

type UserSelection = Array<Record<string, unknown>>;

// Raw JSON state in awareness CRDT
export type RawAwarenessState<
  Flags extends Record<string, unknown> = BlockSuiteFlags,
> = {
  user?: UserInfo;
  color?: string;
  flags: Flags;
  // use v2 to avoid crush on old clients
  selectionV2: Record<string, UserSelection>;
};

export interface AwarenessEvent<
  Flags extends Record<string, unknown> = BlockSuiteFlags,
> {
  id: number;
  type: 'add' | 'update' | 'remove';
  state?: RawAwarenessState<Flags>;
}

export class AwarenessStore<
  Flags extends Record<string, unknown> = BlockSuiteFlags,
> {
  readonly awareness: YAwareness<RawAwarenessState<Flags>>;
  readonly store: Store;

  readonly slots = {
    update: new Slot<AwarenessEvent<Flags>>(),
  };

  constructor(
    store: Store,
    awareness: YAwareness<RawAwarenessState<Flags>>,
    defaultFlags: Flags
  ) {
    this.store = store;
    this.awareness = awareness;
    this.awareness.on('change', this._onAwarenessChange);
    this.awareness.setLocalStateField('selectionV2', {});
    this._initFlags(defaultFlags);
  }

  private _initFlags(defaultFlags: Flags) {
    const upstreamFlags = this.awareness.getLocalState()?.flags;
    const flags = upstreamFlags
      ? merge(true, defaultFlags, upstreamFlags)
      : { ...defaultFlags };
    this.awareness.setLocalStateField('flags', flags);
  }

  setFlag<Key extends keyof Flags>(field: Key, value: Flags[Key]) {
    const oldFlags = this.awareness.getLocalState()?.flags ?? {};
    this.awareness.setLocalStateField('flags', { ...oldFlags, [field]: value });
  }

  getFlag<Key extends keyof Flags>(field: Key): Flags[Key] | undefined {
    const flags = this.awareness.getLocalState()?.flags ?? {};
    return flags[field];
  }

  setReadonly(space: Space, value: boolean): void {
    const flags = this.getFlag('readonly') ?? {};
    this.setFlag('readonly', {
      ...flags,
      [space.id]: value,
    } as Flags['readonly']);
  }

  isReadonly(space: Space): boolean {
    const rd = this.getFlag('readonly');
    if (rd && typeof rd === 'object') {
      return Boolean((rd as Record<string, boolean>)[space.id]);
    } else {
      return false;
    }
  }

  setLocalSelection(space: Space, selection: UserSelection) {
    const oldSelection = this.awareness.getLocalState()?.selectionV2 ?? {};
    this.awareness.setLocalStateField('selectionV2', {
      ...oldSelection,
      [space.id]: selection,
    });
  }

  getLocalSelection(space: Space): ReadonlyArray<Record<string, unknown>> {
    return (this.awareness.getLocalState()?.selectionV2 ?? {})[space.id] ?? [];
  }

  getStates(): Map<number, RawAwarenessState<Flags>> {
    return this.awareness.getStates();
  }

  private _onAwarenessChange = (diff: {
    added: number[];
    removed: number[];
    updated: number[];
  }) => {
    const { added, removed, updated } = diff;

    const states = this.awareness.getStates();
    added.forEach(id => {
      this.slots.update.emit({
        id,
        type: 'add',
        state: states.get(id),
      });
    });
    updated.forEach(id => {
      this.slots.update.emit({
        id,
        type: 'update',
        state: states.get(id),
      });
    });
    removed.forEach(id => {
      this.slots.update.emit({
        id,
        type: 'remove',
      });
    });
  };

  destroy() {
    if (this.awareness) {
      this.awareness.off('change', this._onAwarenessChange);
      this.slots.update.dispose();
    }
  }
}
