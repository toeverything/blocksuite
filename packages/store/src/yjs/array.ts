import type { Array as YArray } from 'yjs';

import type { ProxyConfig } from './config.js';
import { toPlainValue } from './utils.js';

export function subscribeYArray(
  arr: unknown[],
  yArray: YArray<unknown>,
  config: ProxyConfig
): void {
  const { deep = false } = config;
  yArray.observe(event => {
    if (event.changes.keys.size === 0) {
      // skip empty event
      return;
    }
    let retain = 0;
    event.changes.delta.forEach(change => {
      if (change.retain) {
        retain += change.retain;
      }
      if (change.delete) {
        arr.splice(retain, change.delete);
      }
      if (change.insert) {
        if (Array.isArray(change.insert)) {
          const value = deep ? change.insert.map(toPlainValue) : change.insert;
          arr.splice(retain, 0, ...value);
        } else {
          arr.splice(
            retain,
            0,
            deep ? toPlainValue(change.insert) : change.insert
          );
        }
        retain += change.insert.length;
      }
    });
  });
}
