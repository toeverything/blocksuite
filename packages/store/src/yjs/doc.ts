import { debug } from '@blocksuite/global/debug';
import type { Transaction } from 'yjs';
import * as Y from 'yjs';

import type { ProxyConfig } from './config.js';
import { ProxyManager } from './proxy.js';

export type BlockSuiteDocAllowedValue =
  | Record<string, unknown>
  | unknown[]
  | Y.Text;
export type BlockSuiteDocData = Record<string, BlockSuiteDocAllowedValue>;

export class BlockSuiteDoc extends Y.Doc {
  proxy = new ProxyManager();
  private _spaces: Y.Map<Y.Doc> = this.getMap('spaces');

  get spaces() {
    return this._spaces;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  override toJSON(): Record<string, any> {
    const json = super.toJSON();
    delete json.spaces;
    const spaces: Record<string, unknown> = {};
    this.spaces.forEach((doc, key) => {
      spaces[key] = doc.toJSON();
    });
    return {
      ...json,
      spaces,
    };
  }

  getMapProxy<
    Key extends keyof BlockSuiteDocData & string,
    Value extends Record<
      string,
      unknown
    > = BlockSuiteDocData[Key] extends Record<string, unknown>
      ? BlockSuiteDocData[Key]
      : never
  >(key: Key, config: ProxyConfig = {}): Value {
    const map = super.getMap(key);
    return this.proxy.createYProxy(map, config);
  }

  getArrayProxy<
    Key extends keyof BlockSuiteDocData & string,
    Value extends unknown[] = BlockSuiteDocData[Key] extends unknown[]
      ? BlockSuiteDocData[Key]
      : never
  >(key: Key, config: ProxyConfig = {}): Value {
    const array = super.getArray(key);
    return this.proxy.createYProxy(array, config) as Value;
  }

  @debug('transact')
  override transact<T>(f: (arg0: Transaction) => T, origin?: number) {
    return super.transact(f, origin);
  }
}
