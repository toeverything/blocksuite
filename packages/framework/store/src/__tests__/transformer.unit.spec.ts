import { expect, test } from 'vitest';
import * as Y from 'yjs';

import { MemoryBlobManager } from '../adapter/index.js';
import { Text } from '../reactive/index.js';
import {
  defineBlockSchema,
  Schema,
  type SchemaToModel,
} from '../schema/index.js';
import { AssetsManager, BaseBlockTransformer } from '../transformer/index.js';
import { Generator, Workspace } from '../workspace/index.js';

const pageSchema = defineBlockSchema({
  flavour: 'page',
  props: internal => ({
    title: internal.Text('page title'),
    count: 3,
    style: {
      color: 'red',
    },
    items: [
      {
        id: 0,
        content: internal.Text('item 1'),
      },
      {
        id: 1,
        content: internal.Text('item 2'),
      },
      {
        id: 2,
        content: internal.Text('item 3'),
      },
    ],
  }),
  metadata: {
    role: 'root',
    version: 1,
  },
});

type PageBlockModel = SchemaToModel<typeof pageSchema>;

function createTestOptions() {
  const idGenerator = Generator.AutoIncrement;
  const schema = new Schema();
  schema.register([pageSchema]);
  return { id: 'test-workspace', idGenerator, schema };
}

const transformer = new BaseBlockTransformer();
const blobManager = new MemoryBlobManager();
const assets = new AssetsManager({ blob: blobManager });

test('model to snapshot', () => {
  const options = createTestOptions();
  const workspace = new Workspace(options);
  const page = workspace.createPage({ id: 'home' });
  page.load();
  page.addBlock('page');
  const root = page.root as PageBlockModel;

  expect(root).not.toBeNull();
  const snapshot = transformer.toSnapshot({
    model: root,
    assets,
  });
  expect(snapshot).toMatchSnapshot();
});

test('snapshot to model', async () => {
  const options = createTestOptions();
  const workspace = new Workspace(options);
  const page = workspace.createPage({ id: 'home' });
  page.load();
  page.addBlock('page');
  const root = page.root as PageBlockModel;

  const tempDoc = new Y.Doc();
  const map = tempDoc.getMap('temp');

  expect(root).not.toBeNull();
  const snapshot = await transformer.toSnapshot({
    model: root,
    assets,
  });

  const model = await transformer.fromSnapshot({
    json: snapshot,
    assets,
    children: [],
  });
  expect(model.flavour).toBe(root.flavour);

  // @ts-ignore
  expect(model.props.title).toBeInstanceOf(Text);

  // @ts-ignore
  map.set('title', model.props.title.yText);
  // @ts-ignore
  expect(model.props.title.toString()).toBe('page title');

  // @ts-ignore
  expect(model.props.style).toEqual({
    color: 'red',
  });

  // @ts-ignore
  expect(model.props.count).toBe(3);

  // @ts-ignore
  expect(model.props.items).toMatchObject([
    {
      id: 0,
    },
    {
      id: 1,
    },
    {
      id: 2,
    },
  ]);

  // @ts-ignore
  model.props.items.forEach((item, index) => {
    expect(item.content).toBeInstanceOf(Text);
    const key = `item:${index}:content`;
    map.set(key, item.content.yText);
    expect(item.content.toString()).toBe(`item ${index + 1}`);
  });
});
