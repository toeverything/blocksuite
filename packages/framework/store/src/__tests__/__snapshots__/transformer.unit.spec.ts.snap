// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`model to snapshot 1`] = `
{
  "flavour": "page",
  "id": "0",
  "props": {
    "count": 3,
    "items": [
      {
        "content": {
          "$blocksuite:internal:text$": true,
          "delta": [
            {
              "insert": "item 1",
            },
          ],
        },
        "id": 0,
      },
      {
        "content": {
          "$blocksuite:internal:text$": true,
          "delta": [
            {
              "insert": "item 2",
            },
          ],
        },
        "id": 1,
      },
      {
        "content": {
          "$blocksuite:internal:text$": true,
          "delta": [
            {
              "insert": "item 3",
            },
          ],
        },
        "id": 2,
      },
    ],
    "style": {
      "color": "red",
    },
    "title": {
      "$blocksuite:internal:text$": true,
      "delta": [
        {
          "insert": "page title",
        },
      ],
    },
  },
  "version": 1,
}
`;
